### ISSUE TITLE
Memory leaking on return Struct

### ISSUE THREAD

**Issue by penguindark** [vlang/v#7816]
This code can be used to test memory leak on return structs
```v
struct Tst{
mut:
    a int = -1
    b string
}

fn create_data() Tst {
    mut test := Tst{}
    test.a = 1
    return test
}

fn (mut t Tst) fn1(in_txt string){
    t.b = in_txt + "a"
    t.b = t.b + "b"
    t.a++
}

fn main(){
    mut i := 0
    
    for i < 1000000 {
        mut test := create_data()
        test.fn1("ciao $i")
    }
}
```

**Comment by tkshnwesper** [comment]
I'm not sure it was intended, but the `i` variable isn't getting incremented in the `for` loop.

**Comment by JalonSolov** [comment]
Just means it will loop forever, since it will always be < 1000000.

**Comment by tkshnwesper** [comment]
The memory leak is happening while calling the method on the mutable struct. I'm able to reproduce the leak with just this:

```v
struct Tst {
mut:
	a int = -1
	b string
}

fn (mut t Tst) fn1(in_txt string) {
	t.b = in_txt + 'a'
	t.b = t.b + 'b'
	t.a++
}

fn main() {
	mut i := 0
	for i < 10 {
		mut test := Tst{a: 1}
		test.fn1('ciao $i')
		i++
	}
}
```

**Comment by tkshnwesper** [comment]
`create_data()` also seems to leak.

```v
struct Tst {
mut:
	a int = -1
	b string
}

fn create_data() Tst {
    mut test := Tst{}
    test.a = 1
    return test
}

fn main() {
	mut i := 0
	for i < 10 {
		t := create_data()
        println(t.a)
		i++
	}
}
```

**Comment by spytheman** [comment]
adding this method:
```v
[unsafe]
[manualfree]
fn (mut t Tst) free() {
    unsafe { free(t.b.str) }
}
```
... fixes the leak, when the example is compiled with `-autofree`

**Comment by tkshnwesper** [comment]
I narrowed down to where the memory leak is happening with this code:

```v
println(1)
```

It occurs while converting `int` to `str`. See [`int.str_l`](https://github.com/vlang/v/blob/master/vlib/builtin/int.v#L62).

**Comment by ArtemkaKun** [comment]
Hi,
confirmed on Linux.

V version - 0.2.4 (actual for 06.05.2022)
Code compiled with - `v -autofree` and `v -prod -autofree`

### Valgrind log (`valgrind --leak-check-full`)

**`v -autofree`** (run for some time, stopped with Ctrl+C)
```
==174673== Memcheck, a memory error detector
==174673== Copyright (C) 2002-2022, and GNU GPL'd, by Julian Seward et al.
==174673== Using Valgrind-3.19.0 and LibVEX; rerun with -h for copyright info
==174673== Command: ./v-bugs
==174673==
^C==174673==
==174673== Process terminating with default action of signal 2 (SIGINT)
==174673==    at 0x484F790: memset (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==174673==    by 0x434E32: main__main (in /home/yuart/Projects/v-bugs/v-bugs)
==174673==    by 0x447130: main (in /home/yuart/Projects/v-bugs/v-bugs)
==174673==
==174673== HEAP SUMMARY:
==174673==     in use at exit: 74,854,787 bytes in 8,315,301 blocks
==174673==   total heap usage: 33,261,150 allocs, 24,945,849 frees, 2,328,296,480 bytes allocated
==174673==
==174673== 36 bytes in 4 blocks are possibly lost in loss record 5 of 21
==174673==    at 0x4845899: malloc (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==174673==    by 0x41B6FF: malloc_noscan (in /home/yuart/Projects/v-bugs/v-bugs)
==174673==    by 0x42902B: string__plus (in /home/yuart/Projects/v-bugs/v-bugs)
==174673==    by 0x434CDD: main__Tst_fn1 (in /home/yuart/Projects/v-bugs/v-bugs)
==174673==    by 0x434F46: main__main (in /home/yuart/Projects/v-bugs/v-bugs)
==174673==    by 0x447130: main (in /home/yuart/Projects/v-bugs/v-bugs)
==174673==
==174673== 74,837,502 bytes in 8,315,278 blocks are definitely lost in loss record 21 of 21
==174673==    at 0x4845899: malloc (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==174673==    by 0x41B6FF: malloc_noscan (in /home/yuart/Projects/v-bugs/v-bugs)
==174673==    by 0x42902B: string__plus (in /home/yuart/Projects/v-bugs/v-bugs)
==174673==    by 0x434CDD: main__Tst_fn1 (in /home/yuart/Projects/v-bugs/v-bugs)
==174673==    by 0x434F46: main__main (in /home/yuart/Projects/v-bugs/v-bugs)
==174673==    by 0x447130: main (in /home/yuart/Projects/v-bugs/v-bugs)
==174673==
==174673== LEAK SUMMARY:
==174673==    definitely lost: 74,837,502 bytes in 8,315,278 blocks
==174673==    indirectly lost: 0 bytes in 0 blocks
==174673==      possibly lost: 36 bytes in 4 blocks
==174673==    still reachable: 17,249 bytes in 19 blocks
==174673==         suppressed: 0 bytes in 0 blocks
==174673== Reachable blocks (those to which a pointer was found) are not shown.
==174673== To see them, rerun with: --leak-check=full --show-leak-kinds=all
==174673==
==174673== For lists of detected and suppressed errors, rerun with: -s
==174673== ERROR SUMMARY: 2 errors from 2 contexts (suppressed: 0 from 0)
```

**`v -prod -autofree`** (run for some time, stopped with Ctrl+C)
```
==176856== Memcheck, a memory error detector
==176856== Copyright (C) 2002-2022, and GNU GPL'd, by Julian Seward et al.
==176856== Using Valgrind-3.19.0 and LibVEX; rerun with -h for copyright info
==176856== Command: ./v-bugs
==176856== 
^C==176856== 
==176856== Process terminating with default action of signal 2 (SIGINT)
==176856==    at 0x4A08480: __memcpy_avx_unaligned_erms (in /usr/lib/libc.so.6)
==176856==    by 0x110AA3: str_intp.constprop.0 (in /home/yuart/Projects/v-bugs/v-bugs)
==176856==    by 0x1091D5: main (in /home/yuart/Projects/v-bugs/v-bugs)
==176856== 
==176856== HEAP SUMMARY:
==176856==     in use at exit: 74,452,624 bytes in 8,270,589 blocks
==176856==   total heap usage: 33,082,296 allocs, 24,811,707 frees, 2,315,776,823 bytes allocated
==176856== 
==176856== 7 bytes in 1 blocks are definitely lost in loss record 1 of 22
==176856==    at 0x4845899: malloc (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==176856==    by 0x110C41: malloc_noscan (in /home/yuart/Projects/v-bugs/v-bugs)
==176856==    by 0x110A95: str_intp.constprop.0 (in /home/yuart/Projects/v-bugs/v-bugs)
==176856==    by 0x1091D5: main (in /home/yuart/Projects/v-bugs/v-bugs)
==176856==
==176856== 8 bytes in 1 blocks are definitely lost in loss record 2 of 22
==176856==    at 0x4845899: malloc (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==176856==    by 0x115639: _vinit.constprop.0 (in /home/yuart/Projects/v-bugs/v-bugs)
==176856==    by 0x109172: main (in /home/yuart/Projects/v-bugs/v-bugs)
==176856==
==176856== 18 bytes in 2 blocks are possibly lost in loss record 3 of 22
==176856==    at 0x4845899: malloc (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==176856==    by 0x110C41: malloc_noscan (in /home/yuart/Projects/v-bugs/v-bugs)
==176856==    by 0x10921D: main (in /home/yuart/Projects/v-bugs/v-bugs)
==176856==
==176856== 24 bytes in 1 blocks are definitely lost in loss record 4 of 22
==176856==    at 0x484AA83: calloc (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==176856==    by 0x11238D: _vinit.constprop.0 (in /home/yuart/Projects/v-bugs/v-bugs)
==176856==    by 0x109172: main (in /home/yuart/Projects/v-bugs/v-bugs)
==176856==
==176856== 144 bytes in 1 blocks are definitely lost in loss record 8 of 22
==176856==    at 0x484AA83: calloc (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==176856==    by 0x112B9A: _vinit.constprop.0 (in /home/yuart/Projects/v-bugs/v-bugs)
==176856==    by 0x109172: main (in /home/yuart/Projects/v-bugs/v-bugs)
==176856==
==176856== 256 bytes in 1 blocks are definitely lost in loss record 13 of 22
==176856==    at 0x484AA83: calloc (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==176856==    by 0x11249C: _vinit.constprop.0 (in /home/yuart/Projects/v-bugs/v-bugs)
==176856==    by 0x109172: main (in /home/yuart/Projects/v-bugs/v-bugs)
==176856==
==176856== 256 bytes in 1 blocks are definitely lost in loss record 14 of 22
==176856==    at 0x484AA83: calloc (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==176856==    by 0x112525: _vinit.constprop.0 (in /home/yuart/Projects/v-bugs/v-bugs)
==176856==    by 0x109172: main (in /home/yuart/Projects/v-bugs/v-bugs)
==176856==
==176856== 256 bytes in 1 blocks are definitely lost in loss record 15 of 22
==176856==    at 0x484AA83: calloc (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==176856==    by 0x1125AE: _vinit.constprop.0 (in /home/yuart/Projects/v-bugs/v-bugs)
==176856==    by 0x109172: main (in /home/yuart/Projects/v-bugs/v-bugs)
==176856==
==176856== 256 bytes in 1 blocks are definitely lost in loss record 16 of 22
==176856==    at 0x484AA83: calloc (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==176856==    by 0x11268F: _vinit.constprop.0 (in /home/yuart/Projects/v-bugs/v-bugs)
==176856==    by 0x109172: main (in /home/yuart/Projects/v-bugs/v-bugs)
==176856==
==176856== 2,472 bytes in 1 blocks are definitely lost in loss record 18 of 22
==176856==    at 0x484AA83: calloc (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==176856==    by 0x11272F: _vinit.constprop.0 (in /home/yuart/Projects/v-bugs/v-bugs)
==176856==    by 0x109172: main (in /home/yuart/Projects/v-bugs/v-bugs)
==176856==
==176856== 2,592 bytes in 1 blocks are definitely lost in loss record 19 of 22
==176856==    at 0x484AA83: calloc (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==176856==    by 0x112788: _vinit.constprop.0 (in /home/yuart/Projects/v-bugs/v-bugs)
==176856==    by 0x109172: main (in /home/yuart/Projects/v-bugs/v-bugs)
==176856==
==176856== 74,435,103 bytes in 8,270,567 blocks are definitely lost in loss record 22 of 22
==176856==    at 0x4845899: malloc (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==176856==    by 0x110C41: malloc_noscan (in /home/yuart/Projects/v-bugs/v-bugs)
==176856==    by 0x10921D: main (in /home/yuart/Projects/v-bugs/v-bugs)
==176856==
==176856== LEAK SUMMARY:
==176856==    definitely lost: 74,441,374 bytes in 8,270,577 blocks
==176856==    indirectly lost: 0 bytes in 0 blocks
==176856==      possibly lost: 18 bytes in 2 blocks
==176856==    still reachable: 11,232 bytes in 10 blocks
==176856==         suppressed: 0 bytes in 0 blocks
==176856== Reachable blocks (those to which a pointer was found) are not shown.
==176856== To see them, rerun with: --leak-check=full --show-leak-kinds=all
==176856==
==176856== For lists of detected and suppressed errors, rerun with: -s
==176856== ERROR SUMMARY: 12 errors from 12 contexts (suppressed: 0 from 0)
```

**Comment by medvednikov** [comment]
Fixed in 0.3
